README.md
# pub2012memorynets

This is the simulation code which I used to generate data that Richard Naud
plotted in Figure 17.11 of the spiking Hopfield network in Gerstner, W.,
Kistler, W.M., Naud, R., and Paninski, L. (2014). Neuronal dynamics: from
single neurons to networks and models of cognition (Cambridge: Cambridge
University Press).

## Files

It includes
* This file README.md
* The simulation files `sim_hopfield.cpp` and `sim_hopfield2.cpp`
* A `Makefile` to build them
* A `run.sh` file which illstrates the usage of the simulation
* A simple script to visualize output `plot_activity.gnu`
* The patterns I used in the book in the directory `./stim`
* Some additional files `history.txt` and `params_sim2.txt` with some parameter values that worked for me to make the original figure in the book


## Compile

The version I am publishing here has been adapted to compile with Auryn v0.8,
but you find older revisions of the code in the commit history. 

Either way you will need Auryn (fzenke.net/auryn) installed for this to run.
You can build the simulation code with `make` after having updated the
`Makefile` you point to your Auryn library.  


## Example

Here is an example
```
fzenke@grumpy:~/archive/2012-book-wulfram/spiking_hopfield/src$ make && mpirun -n 2 ./sim_hopfield2 --load ../stim/wulfp3.pat --chi 0.3 --gamma 0.2  --wei1 0.15 --wei2 0.1 --simtime 30
make: Nothing to be done for `all'.
load from matrix ./stim/wulfp3.pat.
chi from matrix 0.3.
gamma factor 0.2.
wei1 factor 0.15.
wei2 factor 0.1.
simtime 30.
load from matrix ./stim/wulfp3.pat.
chi from matrix 0.3.
gamma factor 0.2.
wei1 factor 0.15.
wei2 factor 0.1.
simtime 30.
( 0) Setting up neuron groups ...
( 0) Setting up external connections ...
( 0) Setting up EE connections ...
( 0) Setting up EI connections ...
( 0) Setting up IE connections ...
(!!) on rank 1: I1E (Connection.cpp): Some elements of pattern 0 are larger than the underlying NeuronGroups!
(!!) on rank 0: I1E (Connection.cpp): Some elements of pattern 0 are larger than the underlying NeuronGroups!
( 0) Setting up II connections ...
( 0) Setting up stimulus ...
( 0) Setting up monitors ...
( 0) Simulating ...
[=========================] 100%      t=30.0s  f=10.9 Hz in NeuronGroup   
( 0) Freeing ...
```

which will produce a bunch of output files. With `plot_activity.gnu you see how this indeed switches between different
patterns. 
```
fzenke@grumpy:~/archive/2012-book-wulfram/spiking_hopfield/src$ gnuplot 

	G N U P L O T
	Version 5.0 patchlevel 1    last modified 2015-06-07 

	Copyright (C) 1986-1993, 1998, 2004, 2007-2015
	Thomas Williams, Colin Kelley and many others

	gnuplot home:     http://www.gnuplot.info
	faq, bugs, etc:   type "help FAQ"
	immediate help:   type "help"  (plot window: hit 'h')

Terminal type set to 'wxt'
gnuplot> load 'plot_activity.gnu'
```

The output should look somewhat like this.
![Example plot](/images/plot_activity.png)

You can see that there is a pattern switch happening in those 30s.
Firing rates become "less crazy" with paramter tuning and as you load more
patterns (there are files with different numbers of random patterns in
``./stim``) which unfortunately always requires retuning.

## Tuning parameters

`run.sh` gives you an idea of how to run the simulation code. 

Moreover, `./sim_hopfield2 --help` will show you which paramters are exposed via CLI commands.

And `history.txt` as well ass `params_sim2.txt` includes some parameter with
which the code worked. However, There were some changes to Auryn over the
years.One included a bug which made the connecting probability slightly smaller
that was fixed. By using a newer version of Auryn where these problems are
fixed will thus also necessitate to  retuned the critical synaptic strength
parameters. However, the ones in `history.txt` and `params_sim2.txt` might be good starting points.


fzenke, Aug 02, 2016


history.txt
27.02.2013
First success with (auryn commit 5447d3056b7445b4256977f98fce1ca24cde70a9)
make sim_wulf_hopfield && mpirun -n 2 ./sim_wulf_hopfield --load ~/stim/wulfp50.pat --chi 1 --wei1 0.3 --wei2 0.05
* still a lot of synchrony/oscillations but recall seems to start working


Seems to work now with 100 patterns (commit 2d5cc98baab29c4855a38ef1487a755a27c318dd)
make sim_wulf_hopfield && mpirun -n 2 ./sim_wulf_hopfield --load ~/stim/wulfp100.pat --chi 0.4 --gamma 0.1 --wei1 0.3 --wei2 0.02 --ext 7


This is still in overwrite mode for the EE weights ().
make sim_wulf_hopfield && mpirun -n 2 ./sim_wulf_hopfield --load ~/stim/wulfp100.pat --chi 0.3 --gamma 0.04 --wei1 0.31 --wei2 0.02 --ext 5


Status quo now: 90 patterns work fine with auryn commit d6e5941486606c8de421f47ff7eb1454d2fc7795 and the EE weights cut off and 
make sim_wulf_hopfield && mpirun -n 2 ./sim_wulf_hopfield --load ~/stim/wulfp90.pat --chi 0.3 --gamma 0.04 --wei1 0.31 --wei2 0.02 --ext 5


next steps: do not crop the weights and against osciallations maybe reimplement with NMDA only


src/sim_hopfield.cpp
/* 
* Copyright 2014 Friedemann Zenke
*
* This file is part of Auryn, a simulation package for plastic
* spiking neural networks.
* 
* Auryn is free software: you can redistribute it and/or modify
* it under the terms of the GNU General Public License as published by
* the Free Software Foundation, either version 3 of the License, or
* (at your option) any later version.
* 
* Auryn is distributed in the hope that it will be useful,
* but WITHOUT ANY WARRANTY; without even the implied warranty of
* MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
* GNU General Public License for more details.
* 
* You should have received a copy of the GNU General Public License
* along with Auryn.  If not, see <http://www.gnu.org/licenses/>.
*/


#include "auryn.h"

#define NE 8000
#define NI 2000

using namespace auryn;

namespace po = boost::program_options;
namespace mpi = boost::mpi;

int main(int ac,char *av[]) {
	std::string dir = ".";

	std::stringstream oss;
	std::string infilename = "";
	std::string strbuf ;
	std::string msg;

	double w = 0.1;
	double chi = 0.20;
	double gamma = 0.20;
	double ext = 10;

	double wee = 0.;
	double wei1 = w;
	double wei2 = 0.1*w;
	double wie = 1;
	double wii = w;
	double w_ext = 0.1;

	double sparseness_e = 0.1;
	double sparseness_i = 0.1;
	double simtime = 500.;

	int errcode = 0;

	// BEGIN Global stuff
	mpi::environment env(ac, av);
	mpi::communicator world;
	communicator = &world;

	oss << dir  << "/spikehopf." << world.rank() << ".";
	std::string outputfile = oss.str();

	char tmp [255];
	std::stringstream logfile;
	logfile << outputfile << "log";
	logger = new Logger(logfile.str(),world.rank());

	sys = new System(&world);
	// END Global stuff
	

    try {

        po::options_description desc("Allowed options");
        desc.add_options()
            ("help", "produce help message")
            ("load", po::value<std::string>(), "input weight matrix")
            ("chi", po::value<double>(), "chi factor")
            ("ext", po::value<double>(), "ext factor")
            ("gamma", po::value<double>(), "gamma factor")
            ("wee", po::value<double>(), "wee weight")
            ("wei1", po::value<double>(), "wei1 weight")
            ("wei2", po::value<double>(), "wei2 weight")
            ("wie", po::value<double>(), "wie weight")
            ("wii", po::value<double>(), "wii weight")
        ;

        po::variables_map vm;        
        po::store(po::parse_command_line(ac, av, desc), vm);
        po::notify(vm);    

        if (vm.count("help")) {
			std::cout << desc << "\n";
            return 1;
        }

        if (vm.count("load")) {
			std::cout << "load from matrix " 
                 << vm["load"].as<string>() << ".\n";
			infilename = vm["load"].as<string>();
        } 

        if (vm.count("chi")) {
			std::cout << "chi from matrix " 
                 << vm["chi"].as<double>() << ".\n";
			chi = vm["chi"].as<double>();
        } 

        if (vm.count("ext")) {
			std::cout << "ext from matrix " 
                 << vm["ext"].as<double>() << ".\n";
			ext = vm["ext"].as<double>();
        } 

        if (vm.count("gamma")) {
			std::cout << "gamma factor " 
                 << vm["gamma"].as<double>() << ".\n";
			gamma = vm["gamma"].as<double>();
        } 

        if (vm.count("wee")) {
			std::cout << "wee factor " 
                 << vm["wee"].as<double>() << ".\n";
			wee = vm["wee"].as<double>();
        } 

        if (vm.count("wei1")) {
			std::cout << "wei1 factor " 
                 << vm["wei1"].as<double>() << ".\n";
			wei1 = vm["wei1"].as<double>();
        } 

        if (vm.count("wei2")) {
			std::cout << "wei2 factor " 
                 << vm["wei2"].as<double>() << ".\n";
			wei2 = vm["wei2"].as<double>();
        } 

        if (vm.count("wie")) {
			std::cout << "wie factor " 
                 << vm["wie"].as<double>() << ".\n";
			wie = vm["wie"].as<double>();
        } 

        if (vm.count("wii")) {
			std::cout << "wii factor " 
                 << vm["wii"].as<double>() << ".\n";
			wii = vm["wii"].as<double>();
        } 

    }
    catch(std::exception& e) {
		std::cerr << "error: " << e.what() << "\n";
        return 1;
    }
    catch(...) {
		std::cerr << "Exception of unknown type!\n";
    }


	// double primetime = 10;
	logger->msg("Setting up neuron groups ...",PROGRESS,true);
	IFGroup * neurons_e = new IFGroup( NE );
	neurons_e->random_nmda(1,1);

	AIFGroup * neurons_i1 = new AIFGroup( NI/2 );
	neurons_i1->set_tau_mem(10e-3);

	IFGroup * neurons_i2 = new IFGroup( NI/2 );
	neurons_i2->set_tau_mem(10e-3);

	PoissonGroup * poisson = new PoissonGroup( 2000, ext);

	logger->msg("Setting up external connections ...",PROGRESS,true);
	SparseConnection * con_exte = new SparseConnection(poisson,neurons_e,w_ext,sparseness_e,GLUT);
	SparseConnection * con_exti2 = new SparseConnection(poisson,neurons_i2,w_ext,sparseness_i,GABA);

	logger->msg("Setting up EE connections ...",PROGRESS,true);
	SparseConnection * con_ee = new SparseConnection( neurons_e,neurons_e,wee,sparseness_e,GLUT); con_ee->set_name("EE");
	con_ee->load_patterns(infilename, chi, true);
	
	logger->msg("Setting up EI connections ...",PROGRESS,true);
	SparseConnection * con_ei1 = new SparseConnection( neurons_e,neurons_i1,wei1,sparseness_e,GLUT); con_ei1->set_name("EI1");
	SparseConnection * con_ei2 = new SparseConnection( neurons_e,neurons_i2,wei2,sparseness_e,GLUT); con_ei2->set_name("EI2");

	logger->msg("Setting up IE connections ...",PROGRESS,true);
	SparseConnection * con_i1e = new SparseConnection( neurons_i1,neurons_e,0.0,sparseness_i,GABA); con_i1e->set_name("I1E");
	SparseConnection * con_i2e = new SparseConnection( neurons_i2,neurons_e,wie,sparseness_i,GABA); con_i2e->set_name("I2E");
	con_i1e->load_patterns(infilename, gamma);

	logger->msg("Setting up II connections ...",PROGRESS,true);
	SparseConnection * con_i1i1 = new SparseConnection( neurons_i1,neurons_i1,wii,sparseness_i,GABA); con_i1i1->set_name("I1I1");
	SparseConnection * con_i2i2 = new SparseConnection( neurons_i2,neurons_i2,wii,sparseness_i,GABA); con_i2i2->set_name("I2I2");

	logger->msg("Setting up stimulus ...",PROGRESS,true);
	if ( infilename != "" ) {

		std::stringstream filename;
		filename << outputfile << "stimtimes";
		StimulusGroup * stimgroup = new StimulusGroup(1000,infilename,filename.str().c_str(),RANDOM,1);
		stimgroup->set_mean_on_period(1);
		stimgroup->set_mean_off_period(10);
		stimgroup->scale = 300;
		stimgroup->randomintervals=false;
		IdentityConnection * con_stim = new IdentityConnection(stimgroup,neurons_e,1.0,NMDA);

		std::vector<double> dist = stimgroup->get_distribution();
		int r = 4;
		for ( int i = 0 ; i < dist.size() ; ++i ) {
			if ( i < r ) 
				dist[i] = 0.99/r;
			else
				dist[i] = 0.01/(dist.size()-r);
		}
		stimgroup->set_distribution(dist);
		

		filename.str("");
		filename << outputfile << "stimtimes2";
		StimulusGroup * stimgroup2 = new StimulusGroup(1000,infilename,filename.str().c_str(),SEQUENTIAL,1);
		stimgroup2->set_mean_on_period(1);
		stimgroup2->set_mean_off_period(30);
		stimgroup2->scale = 200;
		SparseConnection * con_stim2 = new SparseConnection(stimgroup2,neurons_i2,wei1,sparseness_e);

		filename.str("");
		filename << outputfile << "pact";
		PatternMonitor * patmon = new PatternMonitor(neurons_e,filename.str().c_str(),infilename.c_str(),100);
	}



	msg = "Setting up monitors ...";
	logger->msg(msg,PROGRESS,true);

	std::stringstream filename;
	filename << outputfile << "e.ras";
	SpikeMonitor * smon_e = new SpikeMonitor( neurons_e, filename.str().c_str() );

	filename.str("");
	filename.clear();
	filename << outputfile << "i1.ras";
	SpikeMonitor * smon_i1 = new SpikeMonitor( neurons_i1, filename.str().c_str() );

	filename.str("");
	filename.clear();
	filename << outputfile << "i2.ras";
	SpikeMonitor * smon_i2 = new SpikeMonitor( neurons_i2, filename.str().c_str() );

	RateChecker * chk = new RateChecker( neurons_e , -0.1 , 1000. , 100e-3);

	for (int j = 0; j<1000 ; j++) {
	  neurons_e->tadd(j,5.);
	}

	msg = "Setting up monitors ...";
	logger->msg("Simulating ..." ,PROGRESS,true);
	if (!sys->run(simtime,true)) 
			errcode = 1;

	logger->msg("Freeing ..." ,PROGRESS,true);
	delete sys;

	if (errcode)
		env.abort(errcode);

	return errcode;
}


sim_wulf_hopfield2.cpp
/* 
* Copyright 2012-2016 Friedemann Zenke
*
* This file is a simulation written for Auryn, a simulation package for plastic
* spiking neural networks.
* 
* Auryn is free software: you can redistribute it and/or modify
* it under the terms of the GNU General Public License as published by
* the Free Software Foundation, either version 3 of the License, or
* (at your option) any later version.
* 
* Auryn is distributed in the hope that it will be useful,
* but WITHOUT ANY WARRANTY; without even the implied warranty of
* MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
* GNU General Public License for more details.
* 
* You should have received a copy of the GNU General Public License
* along with Auryn.  If not, see <http://www.gnu.org/licenses/>.
*/

#include <iostream>
#include <iomanip>
#include <stdlib.h>
#include <string>

#include <boost/program_options.hpp>
#include <boost/mpi/environment.hpp>
#include <boost/mpi/communicator.hpp>
#include <boost/mpi.hpp>

#include "System.h"
#include "IFGroup.h"
#include "AIFGroup.h"
#include "PoissonGroup.h"
#include "SparseConnection.h"
#include "SymmetricSTDPConnection.h"
#include "WeightSumMonitor.h"
#include "SpikeMonitor.h"
#include "RateChecker.h"
#include "StimulusGroup.h"
#include "IdentityConnection.h"
#include "PatternMonitor.h"

#define NE 8000
#define NI 2000

using namespace auryn;

namespace po = boost::program_options;
namespace mpi = boost::mpi;

int main(int ac,char *av[]) {
	std::string dir = ".";

	std::stringstream oss;
	std::string infilename = "";
	std::string strbuf ;
	std::string msg;

	double w = 0.1;
	double chi = 0.20;
	double gamma = 0.20;
	double ext = 10;

	double wee = 0.;
	double wei1 = w;
	double wei2 = 0.1*w;
	double wie = 1;
	double wii = 1.0;
	double w_ext = 0.1;

	double sparseness_e = 0.1;
	double sparseness_i = 0.1;
	double simtime = 500.;

	int errcode = 0;

	// BEGIN Global stuff
	mpi::environment env(ac, av);
	mpi::communicator world;
	communicator = &world;

	oss << dir  << "/spikehopf." << world.rank() << ".";
	std::string outputfile = oss.str();

	char tmp [255];
	std::stringstream logfile;
	logfile << outputfile << "log";
	logger = new Logger(logfile.str(),world.rank());

	sys = new System(&world);
	// END Global stuff
	

    try {

        po::options_description desc("Allowed options");
        desc.add_options()
            ("help", "produce help message")
            ("load", po::value<std::string>(), "input weight matrix")
            ("chi", po::value<double>(), "chi factor")
            ("ext", po::value<double>(), "ext factor")
            ("gamma", po::value<double>(), "gamma factor")
            ("wee", po::value<double>(), "wee weight")
            ("wei1", po::value<double>(), "wei1 weight")
            ("wei2", po::value<double>(), "wei2 weight")
            ("wie", po::value<double>(), "wie weight")
            ("wii", po::value<double>(), "wii weight")
        ;

        po::variables_map vm;        
        po::store(po::parse_command_line(ac, av, desc), vm);
        po::notify(vm);    

        if (vm.count("help")) {
			std::cout << desc << "\n";
            return 1;
        }

        if (vm.count("load")) {
			std::cout << "load from matrix " 
                 << vm["load"].as<std::string>() << ".\n";
			infilename = vm["load"].as<std::string>();
        } 

        if (vm.count("chi")) {
			std::cout << "chi from matrix " 
                 << vm["chi"].as<double>() << ".\n";
			chi = vm["chi"].as<double>();
        } 

        if (vm.count("ext")) {
			std::cout << "ext from matrix " 
                 << vm["ext"].as<double>() << ".\n";
			ext = vm["ext"].as<double>();
        } 

        if (vm.count("gamma")) {
			std::cout << "gamma factor " 
                 << vm["gamma"].as<double>() << ".\n";
			gamma = vm["gamma"].as<double>();
        } 

        if (vm.count("wee")) {
			std::cout << "wee factor " 
                 << vm["wee"].as<double>() << ".\n";
			wee = vm["wee"].as<double>();
        } 

        if (vm.count("wei1")) {
			std::cout << "wei1 factor " 
                 << vm["wei1"].as<double>() << ".\n";
			wei1 = vm["wei1"].as<double>();
        } 

        if (vm.count("wei2")) {
			std::cout << "wei2 factor " 
                 << vm["wei2"].as<double>() << ".\n";
			wei2 = vm["wei2"].as<double>();
        } 

        if (vm.count("wie")) {
			std::cout << "wie factor " 
                 << vm["wie"].as<double>() << ".\n";
			wie = vm["wie"].as<double>();
        } 

        if (vm.count("wii")) {
			std::cout << "wii factor " 
                 << vm["wii"].as<double>() << ".\n";
			wii = vm["wii"].as<double>();
        } 

    }
    catch(std::exception& e) {
		std::cerr << "error: " << e.what() << "\n";
        return 1;
    }
    catch(...) {
		std::cerr << "Exception of unknown type!\n";
    }


	// double primetime = 10;
	logger->msg("Setting up neuron groups ...",PROGRESS,true);
	IFGroup * neurons_e = new IFGroup( NE );
	neurons_e->random_nmda(1,1);
	neurons_e->set_ampa_nmda_ratio(0.1);

	AIFGroup * neurons_i1 = new AIFGroup( NI/2 );
	neurons_i1->set_tau_mem(10e-3);
	neurons_i1->set_ampa_nmda_ratio(0.1);

	IFGroup * neurons_i2 = new IFGroup( NI/2 );
	neurons_i2->set_tau_mem(10e-3);
	neurons_i2->set_ampa_nmda_ratio(0.1);

	PoissonGroup * poisson = new PoissonGroup( 1000, ext);

	logger->msg("Setting up external connections ...",PROGRESS,true);
	SparseConnection * con_exte = new SparseConnection(poisson,neurons_e,w_ext,sparseness_e,GLUT);
	SparseConnection * con_exti2 = new SparseConnection(poisson,neurons_i2,w_ext,sparseness_i,GABA);

	logger->msg("Setting up EE connections ...",PROGRESS,true);
	SparseConnection * con_ee = new SparseConnection( neurons_e,neurons_e,wee,sparseness_e,GLUT); con_ee->set_name("EE");
	con_ee->load_patterns(infilename, chi);
	
	logger->msg("Setting up EI connections ...",PROGRESS,true);
	SparseConnection * con_ei1 = new SparseConnection( neurons_e,neurons_i1,wei1,sparseness_e,GLUT); con_ei1->set_name("EI1");
	SparseConnection * con_ei2 = new SparseConnection( neurons_e,neurons_i2,wei2,sparseness_e,GLUT); con_ei2->set_name("EI2");

	logger->msg("Setting up IE connections ...",PROGRESS,true);
	SparseConnection * con_i1e = new SparseConnection( neurons_i1,neurons_e,0.0,sparseness_i,GABA); con_i1e->set_name("I1E");
	SparseConnection * con_i2e = new SparseConnection( neurons_i2,neurons_e,wie,sparseness_i,GABA); con_i2e->set_name("I2E");
	con_i1e->load_patterns(infilename, gamma);

	logger->msg("Setting up II connections ...",PROGRESS,true);
	SparseConnection * con_i1i1 = new SparseConnection( neurons_i1,neurons_i1,wii,sparseness_i,GABA); con_i1i1->set_name("I1I1");
	SparseConnection * con_i2i2 = new SparseConnection( neurons_i2,neurons_i2,wii,sparseness_i,GABA); con_i2i2->set_name("I2I2");

	logger->msg("Setting up stimulus ...",PROGRESS,true);
	if ( infilename != "" ) {

		std::stringstream filename;
		filename << outputfile << "stimtimes";
		StimulusGroup * stimgroup = new StimulusGroup(1000,infilename,filename.str().c_str(),RANDOM);
		stimgroup->set_mean_on_period(1);
		stimgroup->set_mean_off_period(10);
		stimgroup->scale = 300;
		stimgroup->randomintervals=false;
		IdentityConnection * con_stim = new IdentityConnection(stimgroup,neurons_e,1.0,NMDA);

		std::vector<double> dist = stimgroup->get_distribution();
		int r = 3;
		for ( int i = 0 ; i < dist.size() ; ++i ) {
			if ( i < r ) 
				dist[i] = 0.99/r;
			else
				dist[i] = 0.01/(dist.size()-r);
		}
		stimgroup->set_distribution(dist);
		

		filename.str("");
		filename << outputfile << "stimtimes2";
		StimulusGroup * stimgroup2 = new StimulusGroup(1000,infilename,filename.str().c_str(),SEQUENTIAL);
		stimgroup2->set_mean_on_period(1);
		stimgroup2->set_mean_off_period(30);
		stimgroup2->scale = 200;
		SparseConnection * con_stim2 = new SparseConnection(stimgroup2,neurons_i2,0.1,sparseness_e);

		filename.str("");
		filename << outputfile << "pact";
		PatternMonitor * patmon = new PatternMonitor(neurons_e,filename.str().c_str(),infilename.c_str(),100);
	}



	msg = "Setting up monitors ...";
	logger->msg(msg,PROGRESS,true);

	std::stringstream filename;
	filename << outputfile << "e.ras";
	SpikeMonitor * smon_e = new SpikeMonitor( neurons_e, filename.str().c_str() );

	filename.str("");
	filename.clear();
	filename << outputfile << "i1.ras";
	SpikeMonitor * smon_i1 = new SpikeMonitor( neurons_i1, filename.str().c_str() );

	filename.str("");
	filename.clear();
	filename << outputfile << "i2.ras";
	SpikeMonitor * smon_i2 = new SpikeMonitor( neurons_i2, filename.str().c_str() );

	RateChecker * chk = new RateChecker( neurons_e , -0.1 , 1000. , 100e-3);

	for (int j = 0; j<1000 ; j++) {
	  neurons_e->tadd(j,5.);
	}

	msg = "Setting up monitors ...";
	logger->msg("Simulating ..." ,PROGRESS,true);
	if (!sys->run(simtime,true)) 
			errcode = 1;

	logger->msg("Freeing ..." ,PROGRESS,true);
	delete sys;

	if (errcode)
		env.abort(errcode);

	return errcode;
}


src/sim_hopfield2.cpp
/* 
* Copyright 2012-2016 Friedemann Zenke
*
* This file is a simulation written for Auryn, a simulation package for plastic
* spiking neural networks.
* 
* Auryn is free software: you can redistribute it and/or modify
* it under the terms of the GNU General Public License as published by
* the Free Software Foundation, either version 3 of the License, or
* (at your option) any later version.
* 
* Auryn is distributed in the hope that it will be useful,
* but WITHOUT ANY WARRANTY; without even the implied warranty of
* MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
* GNU General Public License for more details.
* 
* You should have received a copy of the GNU General Public License
* along with Auryn.  If not, see <http://www.gnu.org/licenses/>.
*/

#include <iostream>
#include <iomanip>
#include <stdlib.h>
#include <string>

#include <boost/program_options.hpp>
#include <boost/mpi/environment.hpp>
#include <boost/mpi/communicator.hpp>
#include <boost/mpi.hpp>

#include "auryn.h"

#define NE 8000
#define NI 2000

using namespace auryn;

namespace po = boost::program_options;
namespace mpi = boost::mpi;

int main(int ac,char *av[]) {
	std::string dir = ".";

	std::stringstream oss;
	std::string infilename = "";
	std::string strbuf ;
	std::string msg;

	double w = 0.1;
	double chi = 0.20;
	double gamma = 0.20;
	double ext = 10;

	double wee = 0.;
	double wei1 = w;
	double wei2 = 0.1*w;
	double wie = 1;
	double wii = 1.0;
	double w_ext = 0.1;

	double sparseness_e = 0.1;
	double sparseness_i = 0.1;
	double simtime = 500.;

	int errcode = 0;

	// BEGIN Global stuff
	mpi::environment env(ac, av);
	mpi::communicator world;
	communicator = &world;

	oss << dir  << "/spikehopf." << world.rank() << ".";
	std::string outputfile = oss.str();

	char tmp [255];
	std::stringstream logfile;
	logfile << outputfile << "log";
	logger = new Logger(logfile.str(),world.rank());

	sys = new System(&world);
	// END Global stuff
	

    try {

        po::options_description desc("Allowed options");
        desc.add_options()
            ("help", "produce help message")
            ("load", po::value<std::string>(), "input weight matrix")
            ("chi", po::value<double>(), "chi factor")
            ("ext", po::value<double>(), "ext factor")
            ("gamma", po::value<double>(), "gamma factor")
            ("wee", po::value<double>(), "wee weight")
            ("wei1", po::value<double>(), "wei1 weight")
            ("wei2", po::value<double>(), "wei2 weight")
            ("wie", po::value<double>(), "wie weight")
            ("wii", po::value<double>(), "wii weight")
            ("simtime", po::value<double>(), "simulation time")
        ;

        po::variables_map vm;        
        po::store(po::parse_command_line(ac, av, desc), vm);
        po::notify(vm);    

        if (vm.count("help")) {
			std::cout << desc << "\n";
            return 1;
        }

        if (vm.count("load")) {
			std::cout << "load from matrix " 
                 << vm["load"].as<std::string>() << ".\n";
			infilename = vm["load"].as<std::string>();
        } 

        if (vm.count("chi")) {
			std::cout << "chi from matrix " 
                 << vm["chi"].as<double>() << ".\n";
			chi = vm["chi"].as<double>();
        } 

        if (vm.count("ext")) {
			std::cout << "ext from matrix " 
                 << vm["ext"].as<double>() << ".\n";
			ext = vm["ext"].as<double>();
        } 

        if (vm.count("gamma")) {
			std::cout << "gamma factor " 
                 << vm["gamma"].as<double>() << ".\n";
			gamma = vm["gamma"].as<double>();
        } 

        if (vm.count("wee")) {
			std::cout << "wee factor " 
                 << vm["wee"].as<double>() << ".\n";
			wee = vm["wee"].as<double>();
        } 

        if (vm.count("wei1")) {
			std::cout << "wei1 factor " 
                 << vm["wei1"].as<double>() << ".\n";
			wei1 = vm["wei1"].as<double>();
        } 

        if (vm.count("wei2")) {
			std::cout << "wei2 factor " 
                 << vm["wei2"].as<double>() << ".\n";
			wei2 = vm["wei2"].as<double>();
        } 

        if (vm.count("wie")) {
			std::cout << "wie factor " 
                 << vm["wie"].as<double>() << ".\n";
			wie = vm["wie"].as<double>();
        } 

        if (vm.count("wii")) {
			std::cout << "wii factor " 
                 << vm["wii"].as<double>() << ".\n";
			wii = vm["wii"].as<double>();
        } 

        if (vm.count("simtime")) {
			std::cout << "simtime " 
                 << vm["simtime"].as<double>() << ".\n";
			simtime = vm["simtime"].as<double>();
        } 

    }
    catch(std::exception& e) {
		std::cerr << "error: " << e.what() << "\n";
        return 1;
    }
    catch(...) {
		std::cerr << "Exception of unknown type!\n";
    }


	// double primetime = 10;
	logger->msg("Setting up neuron groups ...",PROGRESS,true);
	IFGroup * neurons_e = new IFGroup( NE );
	neurons_e->random_nmda(1,1);
	neurons_e->set_ampa_nmda_ratio(0.1);

	AIFGroup * neurons_i1 = new AIFGroup( NI/2 );
	neurons_i1->set_tau_mem(10e-3);
	neurons_i1->set_ampa_nmda_ratio(0.1);

	IFGroup * neurons_i2 = new IFGroup( NI/2 );
	neurons_i2->set_tau_mem(10e-3);
	neurons_i2->set_ampa_nmda_ratio(0.1);

	PoissonGroup * poisson = new PoissonGroup( 1000, ext);

	logger->msg("Setting up external connections ...",PROGRESS,true);
	SparseConnection * con_exte = new SparseConnection(poisson,neurons_e,w_ext,sparseness_e,GLUT);
	SparseConnection * con_exti2 = new SparseConnection(poisson,neurons_i2,w_ext,sparseness_i,GABA);

	logger->msg("Setting up EE connections ...",PROGRESS,true);
	SparseConnection * con_ee = new SparseConnection( neurons_e,neurons_e,wee,sparseness_e,GLUT); con_ee->set_name("EE");
	con_ee->load_patterns(infilename, chi);
	
	logger->msg("Setting up EI connections ...",PROGRESS,true);
	SparseConnection * con_ei1 = new SparseConnection( neurons_e,neurons_i1,wei1,sparseness_e,GLUT); con_ei1->set_name("EI1");
	SparseConnection * con_ei2 = new SparseConnection( neurons_e,neurons_i2,wei2,sparseness_e,GLUT); con_ei2->set_name("EI2");

	logger->msg("Setting up IE connections ...",PROGRESS,true);
	SparseConnection * con_i1e = new SparseConnection( neurons_i1,neurons_e,0.0,sparseness_i,GABA); con_i1e->set_name("I1E");
	SparseConnection * con_i2e = new SparseConnection( neurons_i2,neurons_e,wie,sparseness_i,GABA); con_i2e->set_name("I2E");
	con_i1e->load_patterns(infilename, gamma);

	logger->msg("Setting up II connections ...",PROGRESS,true);
	SparseConnection * con_i1i1 = new SparseConnection( neurons_i1,neurons_i1,wii,sparseness_i,GABA); con_i1i1->set_name("I1I1");
	SparseConnection * con_i2i2 = new SparseConnection( neurons_i2,neurons_i2,wii,sparseness_i,GABA); con_i2i2->set_name("I2I2");

	logger->msg("Setting up stimulus ...",PROGRESS,true);
	if ( infilename != "" ) {

		std::stringstream filename;
		filename << outputfile << "stimtimes";
		StimulusGroup * stimgroup = new StimulusGroup(1000,infilename,filename.str().c_str(),RANDOM);
		stimgroup->set_mean_on_period(1);
		stimgroup->set_mean_off_period(10);
		stimgroup->scale = 300;
		stimgroup->randomintervals=false;
		IdentityConnection * con_stim = new IdentityConnection(stimgroup,neurons_e,1.0,NMDA);

		std::vector<double> dist = stimgroup->get_distribution();
		int r = 3;
		for ( int i = 0 ; i < dist.size() ; ++i ) {
			if ( i < r ) 
				dist[i] = 0.99/r;
			else
				dist[i] = 0.01/(dist.size()-r);
		}
		stimgroup->set_distribution(dist);
		

		filename.str("");
		filename << outputfile << "stimtimes2";
		StimulusGroup * stimgroup2 = new StimulusGroup(1000,infilename,filename.str().c_str(),SEQUENTIAL);
		stimgroup2->set_mean_on_period(1);
		stimgroup2->set_mean_off_period(30);
		stimgroup2->scale = 200;
		SparseConnection * con_stim2 = new SparseConnection(stimgroup2,neurons_i2,0.1,sparseness_e);

		filename.str("");
		filename << outputfile << "pact";
		PatternMonitor * patmon = new PatternMonitor(neurons_e,filename.str().c_str(),infilename.c_str(),100);
	}



	msg = "Setting up monitors ...";
	logger->msg(msg,PROGRESS,true);

	std::stringstream filename;
	filename << outputfile << "e.ras";
	SpikeMonitor * smon_e = new SpikeMonitor( neurons_e, filename.str().c_str() );

	filename.str("");
	filename.clear();
	filename << outputfile << "i1.ras";
	SpikeMonitor * smon_i1 = new SpikeMonitor( neurons_i1, filename.str().c_str() );

	filename.str("");
	filename.clear();
	filename << outputfile << "i2.ras";
	SpikeMonitor * smon_i2 = new SpikeMonitor( neurons_i2, filename.str().c_str() );

	RateChecker * chk = new RateChecker( neurons_e , -0.1 , 1000. , 100e-3);

	for (int j = 0; j<1000 ; j++) {
	  neurons_e->tadd(j,5.);
	}

	msg = "Setting up monitors ...";
	logger->msg("Simulating ..." ,PROGRESS,true);
	if (!sys->run(simtime,true)) 
			errcode = 1;

	logger->msg("Freeing ..." ,PROGRESS,true);
	delete sys;

	if (errcode)
		env.abort(errcode);

	return errcode;
}
